teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -0,0 +1,213 @@
+- [tektoncd/pipeline#1486](https://github.com/tektoncd/pipeline/pull/1486)
+- [tektoncd/pipeline#1429](https://github.com/tektoncd/pipeline/pull/1429)
+- [tektoncd/experimental#479](https://github.com/tektoncd/experimental/issues/479)
+- [tektoncd/pipeline#2452](https://github.com/tektoncd/pipeline/issues/2452)
+
+### Goals
+
+- Define a mechanism for cleaning up completed Runs to minimize resources the
+  Pipeline controller is responsible for.
+- Preserve Run metadata so that users can continue to query it beyond the
+  lifetime of the Pipeline controller.
+
+### Non-Goals
+
+- Providing a programmatic mechanism for end users to dynamically control deletion behavior.
+
+### Use Cases (optional)
+
+This is a commonly requested feature due to the strain Run history this puts on
+etcd over time. See [Relevant Issues under Motivation](#motivation).
+
+## Requirements
+
+- Run metadata should be available even after deletion from the Pipeline
+  controller data plane.
+- Runs should be deleted as soon as it can be done safely (i.e. we have
+  confirmed the data is stored durably elsewhere).
+- Deletion should occur with some configurable delay to allow other integrations
+  watching Runs to respond to events.
+- Operators should be able to opt in / out to deletion behavior.
+
+## Proposal
+
+### Delete Runs when complete
+
+The
+[Tekton Results watcher](https://github.com/tektoncd/results/blob/main/docs/watcher/README.md)
+already watches for changes to TaskRun and PipelineRuns to upload Records for
+long term storage. Once we have verified that 1)
+[the Run has completed](https://github.com/tektoncd/pipeline/blob/master/docs/runs.md#monitoring-execution-status)
+and 2) the Run Record has been successfully stored, we know we can delete the
+Run without affecting execution.
+
+### Grace Period
+
+Tekton Results may not be the only controller listening to changes in TaskRun
+and PipelineRun CRDs (e.g. Notifications that happen post-Run execution).
+Because of this, deleting the Run immediately might cause unintended race
+conditions and affect other integrations using a similar mechanism.
+
+To address this, we should have a user configurable grace period for Run
+deletion, and require that the time elapsed since the Record's last update time
+(since the k8s ObjectMeta does not have a similar concept of last update time)
+is greater than the configured grace period before deletion. If the grace period
+has not been met yet, then we should simply
+[re-enqueue](https://pkg.go.dev/knative.dev/pkg@v0.0.0-20210208175252-a02dcff9ee26/controller#Impl.EnqueueKeyAfter)
+the Run to be handled by the controller at a later point in time.
+
+To start, this configuration will be set by the operator of the Results Watcher via
+a flag. We may consider adding additional end user controls for this behavior
+later (see [Future Work](#grace-period-api-configuration)), but this is out of
+scope for this initial work.
+
+### Notes/Caveats (optional)
```

Comment:
maybe a caveat about how this will allow for the re-creation of a taskrun/pipelinerun with the same name?

and maybe something about the underlying pod being deleted too? (tho we already do that when you cancel 😅 )

Reply:
> Run recreation resulting in independent Results is something that's supported independent of Run cleanup (e.g. you can create/delete/create today and get different results), so I'm not 100% certain if it makes sense w.r.t. this propos

Ah I see what you mean: without this proposal, if you delete a run and recreate it, you get duplicate results, so this proposal isnt introducing the possibility of duplicate results, it's already there.

I think what this proposal IS adding is that if folks were never doing any kind of cleanup on their own (maybe even were preventing it?) they would never see that possibility - but if they enable the functionality described here, they would. I think it might be worth mentioning but maybe not that important given what you pointed out!

> Re pod deletion - is your concern we're no longer preserving pod metadata stored as well? That might be an interesting FR (and one that shouldn't be too hard to add)!

Oh I hadn't thought of that, that's a good point!

I was just thinking that the pod seems like a pretty fundamental type for k8s and i could see a lot of tools being used to investigating pods; if we delete them then they can't (I dont have any actual examples of this, just waving my hands...) (maybe one - not very good - example is even in the GCP console you can view pods - tho actually im not sure you can view pods that arent running so this example is quickly falling apart...)